ext/strscan/strscan.c

```diff
@@ -686,14 +686,6 @@ strscan_do_scan(VALUE self, VALUE pattern, int succptr, int getstr, int headonly
 {
     struct strscanner *p;
 
-    if (headonly) {
-        if (!RB_TYPE_P(pattern, T_REGEXP)) {
-            StringValue(pattern);
-        }
```

Comment:
Could you keep this to report an error for an invalid `pattern` (e.g.: `1`) and `S_RESTLEN(p) < 0` case?

Reply:
https://github.com/ruby/strscan/blob/377d6b232451f9482a7f147b5c311719e9ffd10b/ext/strscan/strscan.c#L685-L694

`S_RESTLEN(p) < 0` is checked here.

https://github.com/ruby/strscan/blob/377d6b232451f9482a7f147b5c311719e9ffd10b/ext/strscan/strscan.c#L708-L710

`StringValue(pattern)` is checked here.

Reply:
I thought `ruby -I lib -r strscan -e 's = StringScanner.new("a"); s.scan("a"); s.scan(1)'` causes the condition but it seems that `S_RESTLEN(p) < 0` will be never happen.